sympy/polys/matrices/ddm.py

```diff
@@ -598,22 +598,66 @@ def rref_den(a):
         return b, denom, pivots
 
     def nullspace(a):
+        """Returns a basis for the nullspace of a.
+
+        The domain of the matrix must be a field.
+
+        See Also
+        ========
+
+        rref
+        sympy.polys.matrices.domainmatrix.DomainMatrix.nullspace
+        """
         rref, pivots = a.rref()
-        rows, cols = a.shape
-        domain = a.domain
+        return rref.nullspace_from_rref(pivots)
+
+    def nullspace_from_rref(a, pivots=None):
+        """Compute the nullspace of a matrix from its rref.
+
+        The domain of the matrix can be any domain.
+
+        Returns a tuple (basis, nonpivots).
+
+        See Also
+        ========
+
+        sympy.polys.matrices.domainmatrix.DomainMatrix.nullspace
+            The higher level interface to this function.
+        """
+        m, n = a.shape
+        K = a.domain
+
+        if pivots is None:
+            pivots = []
+            last_pivot = -1
+            for i in range(m):
+                for j in range(last_pivot+1, n):
+                    aij = a[i][j]
+                    if aij:
```

Comment:
```suggestion
                    if a[i][j]:  # aij
```

Reply:
Good point. The intention was to do `ai = a[i]` outside the loop so that in the loop we only need `ai[j]`. The cost of indexing into a list is small but nonzero so it is still good to minimise repeated indexing in loops.

I've changed this.